In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing import OneHotEncoder,MultiLabelBinarizer,LabelEncoder,MinMaxScaler,StandardScaler, RobustScaler
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import ADASYN
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense,Dropout
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.models import load_model
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
import numpy as np

In [2]:
data=pd.read_csv('Lending_Loan_Company_Data_Analysis.csv')
data_num=data.select_dtypes(exclude='object')

# Feature Engineering

## Check for Collinearity B/W the Columns

In [3]:
for i in data_num.columns:
    corr=data_num.corr()
    corr[i].sort_values(ascending=False)
    print(corr)

                           Credit_Policy  Interest_Rate  Monthly_Installment  \
Credit_Policy                   1.000000      -0.294089             0.058770   
Interest_Rate                  -0.294089       1.000000             0.276140   
Monthly_Installment             0.058770       0.276140             1.000000   
Natural_log_Annual_Income       0.034906       0.056383             0.448102   
Debt_Income_Ratio              -0.090901       0.220006             0.050202   
FICO_Score                      0.348319      -0.714821             0.086039   
Days_with_Credit_Line           0.099026      -0.124022             0.183297   
Revolving_Balance              -0.187518       0.092527             0.233625   
Revolving_Utilization          -0.104095       0.464837             0.081356   
Enquiries_Last_6_Months        -0.535511       0.202780            -0.010419   
Deliquency_Last_2_Yrs          -0.076318       0.156079            -0.004368   
Public_Derogatory_Records      -0.054243

### Since, Most of the Columns have the Scores less Than 0.5 between them, There is no need to drop off the columns

# Model Building

## 1. Encode the Labels

In [4]:
data_model=data.copy()

In [5]:
le=LabelEncoder()

In [6]:
data_model

,Credit_Policy,Loan_Purpose,Interest_Rate,Monthly_Installment,Natural_log_Annual_Income,Debt_Income_Ratio,FICO_Score,Days_with_Credit_Line,Revolving_Balance,Revolving_Utilization,Enquiries_Last_6_Months,Deliquency_Last_2_Yrs,Public_Derogatory_Records,Paid_Status
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,Paid
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,Paid
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,Paid
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,Paid
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,Paid
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9573,0,all_other,0.1461,344.76,12.180755,10.39,672,10474.000000,215372,82.1,2,0,0,Unpaid
9574,0,all_other,0.1253,257.70,11.141862,0.21,722,4380.000000,184,1.1,5,0,0,Unpaid
9575,0,debt_consolidation,0.1071,97.81,10.596635,13.09,687,3450.041667,10036,82.9,8,0,0,Unpaid
9576,0,home_improvement,0.1600,351.58,10.819778,19.18,692,1800.000000,0,3.2,5,0,0,Unpaid


In [7]:
data['Loan_Purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [8]:
data_model['Loan_Purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [9]:
data

,Credit_Policy,Loan_Purpose,Interest_Rate,Monthly_Installment,Natural_log_Annual_Income,Debt_Income_Ratio,FICO_Score,Days_with_Credit_Line,Revolving_Balance,Revolving_Utilization,Enquiries_Last_6_Months,Deliquency_Last_2_Yrs,Public_Derogatory_Records,Paid_Status
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,Paid
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,Paid
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,Paid
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,Paid
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,Paid
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9573,0,all_other,0.1461,344.76,12.180755,10.39,672,10474.000000,215372,82.1,2,0,0,Unpaid
9574,0,all_other,0.1253,257.70,11.141862,0.21,722,4380.000000,184,1.1,5,0,0,Unpaid
9575,0,debt_consolidation,0.1071,97.81,10.596635,13.09,687,3450.041667,10036,82.9,8,0,0,Unpaid
9576,0,home_improvement,0.1600,351.58,10.819778,19.18,692,1800.000000,0,3.2,5,0,0,Unpaid


## Description of Labels

### 1. Loan Purpose
#### * debt_consolidation - 2
#### * credit_card - 1
#### * all_other - 0
#### * home_improvement -4
#### * small_business - 6
#### * major_purchase - 5
#### * educational - 3

### 2. Paid Status
#### * Paid - 0
#### * Unpaid - 1

In [10]:
data_model['Installment_to_Income'] = data_model['Monthly_Installment'] / (data['Natural_log_Annual_Income'] / 12)
data_model['Interest_Installment'] = data_model['Interest_Rate'] * data_model['Monthly_Installment']
data_model['Revolving_to_Income'] = data_model['Revolving_Balance'] / data_model['Natural_log_Annual_Income']
data_model['Credit_Length_Years'] = data_model['Days_with_Credit_Line'] / 365
data_model['Risk_Score'] = data_model['Interest_Rate'] / data_model['FICO_Score']

In [11]:
x=data_model.drop(['Paid_Status','Natural_log_Annual_Income'],axis=1)
y=le.fit_transform(data_model['Paid_Status'])

In [12]:
xtrain,xtemp,ytrain,ytemp=train_test_split(x,y,test_size=0.3,random_state=42)
xeval,xtest,yeval,ytest=train_test_split(xtemp,ytemp,test_size=0.5,random_state=42)

In [13]:

cat_cols = ['Credit_Policy', 'Loan_Purpose', 'Enquiries_Last_6_Months' ,'Deliquency_Last_2_Yrs','Public_Derogatory_Records']

# Outlier numeric
robust_cols = [
    'Interest_Rate',
    'Monthly_Installment',
    'FICO_Score',
    'Days_with_Credit_Line',
    'Revolving_Balance',
    'Installment_to_Income',
    'Interest_Installment'
    
]

# Remaining numeric
standard_cols = [col for col in xtrain.columns 
                 if col not in robust_cols + cat_cols]



In [14]:

preprocessor = ColumnTransformer(
    transformers=[
        ('robust', RobustScaler(), robust_cols),
       
        ('cat', OneHotEncoder(handle_unknown='ignore'), cat_cols)   
    ]
)

# Step 1: Fit scaler on original train
X_train_scaled = preprocessor.fit_transform(xtrain)

# Step 2: Apply same transform
X_eval = preprocessor.transform(xeval)
X_test = preprocessor.transform(xtest)

# Step 3: Apply ADASYN AFTER scaling
adasyn = ADASYN(sampling_strategy=0.5, random_state=42)
X_train_res_Ada, y_train_res = adasyn.fit_resample(X_train_scaled, ytrain)

In [15]:
model = Sequential()

# Input Layer + Hidden Layer 1
model.add(Dense(64, activation='relu', input_dim=X_train_res_Ada.shape[1]))
model.add(Dropout(0.3))

# Hidden Layer 2
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))

# Hidden Layer 3
model.add(Dense(16, activation='relu'))

# Output Layer
model.add(Dense(1, activation='sigmoid'))  # Binary classification

# Compile
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['Precision', 'Recall']
)

# Early Stopping (VERY IMPORTANT)
early_stop = EarlyStopping(
    monitor='val_loss',
    patience=20,
    restore_best_weights=True
)

# Train
history = model.fit(
    X_train_res_Ada, y_train_res,
    validation_data=(X_eval, yeval),
    
    epochs=200,
    batch_size=32,
    callbacks=[early_stop]
    
)

Epoch 1/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 13s 18ms/step - Precision: 0.4068 - Recall: 0.0868 - loss: 0.6296 - val_Precision: 0.3623 - val_Recall: 0.1064 - val_loss: 0.4987
Epoch 2/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 11ms/step - Precision: 0.5455 - Recall: 0.1070 - loss: 0.6119 - val_Precision: 0.4516 - val_Recall: 0.0596 - val_loss: 0.4986
Epoch 3/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - Precision: 0.5500 - Recall: 0.1132 - loss: 0.6070 - val_Precision: 0.4118 - val_Recall: 0.1191 - val_loss: 0.4935
Epoch 4/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - Precision: 0.5836 - Recall: 0.1365 - loss: 0.6013 - val_Precision: 0.4028 - val_Recall: 0.1234 - val_loss: 0.4900
Epoch 5/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 10ms/step - Precision: 0.5757 - Recall: 0.1526 - loss: 0.5991 - val_Precision: 0.3333 - val_Recall: 0.1617 - val_loss: 0.5003
Epoch 6/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 10ms/step - Precision: 0.6094 - Recall: 0.1787 - loss: 0.5921 - val_Precision: 0.3292 - val_Recall: 0.2255 

In [16]:
loss, precision, recall = model.evaluate(X_eval, yeval)
print("Loss:", loss)
print("Precision:", precision)
print("Recall:", recall)

45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - Precision: 0.4028 - Recall: 0.1234 - loss: 0.4900
Loss: 0.4900009036064148
Precision: 0.4027777910232544
Recall: 0.12340425699949265


In [25]:
ypred=model.predict(X_eval)

ypred_t = (model.predict(X_eval) > 0.38).astype(int)

print(classification_report(yeval, ypred_t))


45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step
              precision    recall  f1-score   support

           0       0.88      0.67      0.76      1202
           1       0.25      0.55      0.34       235

    accuracy                           0.65      1437
   macro avg       0.56      0.61      0.55      1437
weighted avg       0.78      0.65      0.69      1437



In [26]:
ypred_Test=model.predict(X_test)

ypred_test = (model.predict(X_test) > 0.38).astype(int)

print(classification_report(ytest, ypred_test))

45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step
45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 19ms/step
              precision    recall  f1-score   support

           0       0.88      0.66      0.75      1206
           1       0.23      0.54      0.32       231

    accuracy                           0.64      1437
   macro avg       0.56      0.60      0.54      1437
weighted avg       0.78      0.64      0.69      1437



# Insights

### 1. Handling Class Imbalance Using ADASYN

The Lending Club dataset contained a significant class imbalance, where ‘Unpaid’ (loan default) cases were much fewer compared to ‘Paid’ loans.

To address this issue, ADASYN (Adaptive Synthetic Sampling) was applied to generate synthetic minority-class samples dynamically based on difficult-to-learn regions of the dataset.

The objective of ADASYN was to improve the ANN model’s ability to identify risky borrowers by exposing the model to more minority-class examples during training.

### 2. Model Performance

The ANN model trained using ADASYN achieved approximately:

Recall ≈ 54–55% for the ‘Unpaid’ class
Precision ≈ 23–25%
F1-score ≈ 0.32–0.34
Accuracy ≈ 64–65%

The model successfully improved the detection of loan defaulters compared to a standard imbalanced ANN model.

However, because ADASYN generates synthetic minority samples, the model became more sensitive toward predicting defaults, leading to a large number of false positives.

This resulted in relatively low precision, meaning that several safe borrowers were incorrectly classified as risky applicants.

### 3. Threshold Tuning

The default classification threshold of 0.5 was not optimal for detecting minority-class borrowers effectively.

To improve recall for the ‘Unpaid’ class, threshold tuning was performed using a threshold value of 0.38.

Effects of lowering the threshold:

Increased recall for defaulters
Improved detection of risky borrowers
Increased false positives
Reduced precision

The threshold of 0.38 provided a practical trade-off between sensitivity and prediction stability for the loan default prediction problem.

### 4. Generalization Performance

The evaluation and test datasets produced very similar results:

Evaluation Dataset
Recall ≈ 55%
Precision ≈ 25%
Test Dataset
Recall ≈ 54%
Precision ≈ 23%

The consistency between evaluation and test performance indicates that the model generalized reasonably well and did not suffer from significant overfitting.

### 5. Business Interpretation

The primary objective of the project was to identify potentially risky loan applicants and reduce losses caused by loan defaults.

Since failing to detect a defaulter can lead to major financial loss, the model was optimized toward improving recall for the minority class.

The ADASYN-based ANN model therefore acts as a risk-screening system that prioritizes the identification of potentially defaulting borrowers, even at the cost of increased false positives.

# Summary

From a business perspective, the ADASYN-based ANN model focused on improving the detection of loan defaulters (‘Unpaid’ customers), which is critical in reducing lending risk.

To improve minority-class detection, ADASYN generated synthetic defaulter samples and the classification threshold was reduced to 0.38. This increased the model’s ability to identify risky borrowers and improved recall for the minority class.

Although the model produced a considerable number of false positives, this trade-off is acceptable in financial risk management because missing a true defaulter is generally more costly than incorrectly flagging a safe borrower.

Overall, the ADASYN-based ANN model served as an effective imbalance-handling approach for improving risky borrower detection while maintaining stable performance across evaluation and test datasets.